src/Tools/dotnet-format/CodeFormatter.cs

```diff
@@ -0,0 +1,131 @@
+    internal class CodeFormatter
+    {
+        public static async Task<int> FormatWorkspaceAsync(ILogger logger, string workspacePath, bool isSolution, CancellationToken token)
+        {
+            logger.LogInformation(string.Format(Resources.Info_FormattingFilesInWorkspace, Path.GetFileName(workspacePath)));
+
+            logger.LogTrace($"Loading workspace.");
+
+            var stopwatch = Stopwatch.StartNew();
+
+            var properties = new Dictionary<string, string>
+            {
+                // This property ensures that XAML files will be compiled in the current AppDomain
+                // rather than a separate one. Any tasks isolated in AppDomains or tasks that create
+                // AppDomains will likely not work due to https://github.com/Microsoft/MSBuildLocator/issues/16.
+                { "AlwaysCompileMarkupFilesInSeparateDomain", bool.FalseString },
+
+                // Use the latest language version to force the full set of available analyzers to run on the project.
+                { "LangVersion", "latest" },
+            };
+            using (var workspace = MSBuildWorkspace.Create(properties))
+            {
+                var projectPath = string.Empty;
+                if (isSolution)
+                {
+                    await workspace.OpenSolutionAsync(workspacePath, cancellationToken: token).ConfigureAwait(false);
+                }
+                else
+                {
+                    await workspace.OpenProjectAsync(workspacePath, cancellationToken: token).ConfigureAwait(false);
+                    projectPath = workspacePath;
+                }
+
+                logger.LogTrace($"Workspace loaded in {stopwatch.ElapsedMilliseconds}ms.");
+                stopwatch.Restart();
+
+                await FormatFilesInSolutionAsync(logger, workspace.CurrentSolution, projectPath, token).ConfigureAwait(false);
+
+                logger.LogTrace($"Format complete in {stopwatch.ElapsedMilliseconds}ms.");
+            }
+
+            logger.LogInformation(Resources.Info_FormatComplete);
+
+            return 0;
+        }
+
+        private static async Task FormatFilesInSolutionAsync(ILogger logger, Solution solution, string projectPath, CancellationToken cancellationToken)
+        {
+            foreach (var projectId in solution.ProjectIds)
+            {
+                var project = solution.GetProject(projectId);
+                if (!string.IsNullOrEmpty(projectPath) && !project.FilePath.Equals(projectPath, StringComparison.OrdinalIgnoreCase))
+                {
+                    continue;
+                }
+
+                logger.LogInformation(string.Format(Resources.Info_FormattingFilesInProject, project.Name));
+
+                if (project.Language != LanguageNames.CSharp && project.Language != LanguageNames.VisualBasic)
+                {
+                    logger.LogWarning(string.Format(Resources.Warn_UnsupportedProject, Path.GetFileName(project.FilePath)));
```

Comment:
Can you pleae make your resource strings match the english text that will be printed.  Otherwise, it's hard to read this and know what the message will be.  In this case the resource ID would be:

Could_not_format_0_Format_currently_supports_only_CSharp_and_Visual_Basic_projects

This is also the IDE style.